core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java

```diff
@@ -17,87 +17,150 @@
     int limit = Math.min(errors.size(), 3);
-    String details =
-        Joiner.on(", ").withKeyValueSeparator(": ").join(Iterables.limit(errors.entrySet(), limit));
-
+    Iterator<Entry<Node, List<Throwable>>> iterator = errors.iterator();
+    StringBuilder details = new StringBuilder();
+    while (iterator.hasNext()) {
+      Entry<Node, List<Throwable>> entry = iterator.next();
+      details.append(entry.getKey()).append(": ").append(entry.getValue());
+      if (iterator.hasNext()) {
+        details.append(", ");
+      }
+    }
     return String.format(
-        baseMessage + " (showing first %d, use getErrors() for more: %s)", limit, details);
+        "%s (showing first %d, use getAllErrors() for more): %s", baseMessage, limit, details);
   }
 
-  /** The details of the individual error on each node. */
+  /**
+   * An immutable map containing the first error on each tried node.
+   *
+   * @deprecated Use {@link #getAllErrors()} instead.
+   */
   @NonNull
+  @Deprecated
   public Map<Node, Throwable> getErrors() {
+    ImmutableMap.Builder<Node, Throwable> builder = ImmutableMap.builder();
+    for (Node node : errors.keySet()) {
+      List<Throwable> nodeErrors = errors.get(node);
+      if (!nodeErrors.isEmpty()) {
+        builder.put(node, nodeErrors.get(0));
+      }
+    }
+    return builder.build();
+  }
+
+  /** An immutable map containing all errors on each tried node. */
+  @NonNull
+  public Map<Node, List<Throwable>> getAllErrors() {
     return errors;
   }
 
   @NonNull
   @Override
   public DriverException copy() {
-    return new AllNodesFailedException(getMessage(), getExecutionInfo(), errors);
+    return new AllNodesFailedException(getMessage(), getExecutionInfo(), errors.entrySet());
   }
 
   @NonNull
   public AllNodesFailedException reword(String newMessage) {
     return new AllNodesFailedException(
-        buildMessage(newMessage, errors), getExecutionInfo(), errors);
+        buildMessage(newMessage, errors.entrySet()), getExecutionInfo(), errors.entrySet());
+  }
+
+  private static Set<Entry<Node, List<Throwable>>> groupByNode(
+      Iterable<Entry<Node, Throwable>> errors) {
+    // no need for immutable collections here
+    Map<Node, List<Throwable>> map = new LinkedHashMap<>();
+    for (Entry<Node, Throwable> entry : errors) {
+      Node node = entry.getKey();
+      Throwable error = entry.getValue();
+      map.compute(
+          node,
+          (k, v) -> {
+            if (v == null) {
+              v = new ArrayList<>();
+            }
+            v.add(error);
+            return v;
+          });
+    }
+    return map.entrySet();
+  }
+
+  private static Map<Node, List<Throwable>> toDeepImmutableMap(
+      Iterable<Entry<Node, List<Throwable>>> errors) {
+    ImmutableMap.Builder<Node, List<Throwable>> builder = ImmutableMap.builder();
+    for (Entry<Node, List<Throwable>> entry : errors) {
+      builder.put(entry.getKey(), ImmutableList.copyOf(entry.getValue()));
```

Comment:
If that is a deep copy, shouldn't we make a safe copy of `entry.getKey()`? Or is the Node returned by `getKey` immutable?

Reply:
`DefaultNode` is mutable, but I don't want to change it in this PR because it was already mutable in the previous implementation of this class. I'm only making the values immutable.

Reply:
`DefaultNode` is mutable but it uses reference equality. Changing the mutable fields will never affect how two instances compare.